src/routes/authorize.rs

```diff
@@ -33,49 +39,99 @@ async fn authorize_handler(
 ) -> Result<HttpResponse, Error> {
     log::info!("GET authorize handler");
     let app_state = app_state.read().await;
-    let redirect_urls: &Vec<url::Url> = &app_state
-        .client_configs
-        .get(&query.client_id)
-        .ok_or(Error::OauthInvalidClientId)?
-        .redirect_urls;
 
-    let requirements_empty = &app_state
-        .client_configs
-        .get(&query.client_id)
-        .ok_or(Error::OauthInvalidClientId)?
-        .requirements
-        .is_empty();
+    // Return an error without redirecting, if the `redirect_uri` is an invalid URL.
+    let redirect_uri =
+        Url::parse(&query.redirect_uri).map_err(|_| Error::OauthInvalidRedirectUri)?;
+
+    // `client_id` must be present.
+    let client_id: String = if let Some(client_id) = &query.client_id {
+        client_id.clone()
+    } else {
+        return error_redirect(
+            redirect_uri,
+            Error::OauthInvalidClientId,
+            query.state.as_deref(),
+        );
+    };
+
+    // `response_type` is required.
+    let response_type: &str = if let Some(response_type) = &query.response_type {
+        response_type
+    } else {
+        return error_redirect(redirect_uri, Error::ResponseType, query.state.as_deref());
+    };
+
+    let client_configs = if let Some(configs) = app_state.client_configs.get(&client_id) {
+        configs
+    } else {
+        return error_redirect(
+            Url::parse(&query.redirect_uri).unwrap(),
+            Error::OauthInvalidClientId,
+            query.state.as_deref(),
+        );
+    };
+
+    let redirect_urls: &Vec<url::Url> = &client_configs.redirect_urls;
+    let requirements_empty = client_configs.requirements.is_empty();
 
     // Support Authorization Code Flow and Implicit Flow.
-    let response_type = ResponseType::from_str(query.response_type.as_str())?;
+    let response_type = ResponseType::from_str(response_type)?;
 
     // Implicit flow must include a nonce.
     if response_type.is_implicit_flow() && query.nonce.is_none() {
-        return Err(Error::InvalidNonce);
+        return error_redirect(redirect_uri, Error::InvalidNonce, query.state.as_deref());
     }
 
-    let is_redirect_uri_in_query = redirect_urls
-        .contains(&Url::parse(&query.redirect_uri).map_err(|_| Error::OauthInvalidRedirectUri)?);
+    let is_redirect_uri_in_query = redirect_urls.contains(&redirect_uri);
     if !is_redirect_uri_in_query {
-        return Err(Error::OauthInvalidRedirectUri);
+        return error_redirect(
+            redirect_uri,
+            Error::OauthInvalidRedirectUri,
+            query.state.as_deref(),
+        );
     }
 
-    session.insert(REDIRECT_URI_SESSION_KEY, query.redirect_uri.clone())?;
-    session.insert(RESPONSE_TYPE_SESSION_KEY, query.response_type.clone())?;
+    let authorize_parameters = AuthorizeParameters {
```

Comment:
Ok so I understand this is more a `session_authorization_parameter` or something like that? Just trying to figure out if it would help renaming the type and the variable to something easier to understand and to distinguish from the existing `AuthorizeQueryParameters`.

Reply:
7e78c10ef9d6